Create migrate-connector-to-poetry.md (#36965)
airbytehq/airbyte: docs/contributing-to-airbyte/migrate-connector-to-poetry.md

# Migrate connector to use Poetry and base Docker Image

1. Remove the `Dockerfile` and `.dockerignore`

> 2. Include the `baseImage` to the `metadata.yaml` file. For now, look for `baseImage` to find examples:

Reviewer: ```suggestion
2. Include the `baseImage` to the `metadata.yaml` file. Please use the [latest non RC](https://hub.docker.com/r/airbyte/python-connector-base/tags) `airbyte/python-connector-base` image 
``` and add the image digest identifier (`@shar456:<digest>`)

> 2. Include the `baseImage` to the `metadata.yaml` file. For now, look for `baseImage` to find examples:

Reviewer: Users can use `airbyte-ci connectors --name=<connector-name> migrate_to_base_image`

```yaml
data:
  allowedHosts:
    hosts:
      - *
  connectorBuildOptions:
    baseImage: docker.io/airbyte/python-connector-base:1.2.0@sha256:c22a9d97464b69d6ef01898edf3f8612dc11614f05a84984451dde195f337db9
  connectorSubtype: api
  connectorType: source
  dockerImageTag: 0.1.0
```

> 3. Create a file called `pyproject.toml` and transfer all lib used in `setup.py` to this file

Reviewer 2: Can we put this into a two-column before/after table instead so it's easy to look at visually? Or tabs? 

If we use either of those, will it look good on mobile layout, and will it revert to vertical layout automatically?

> 3. Create a file called `pyproject.toml` and transfer all lib used in `setup.py` to this file

Reviewer: Users could also try `airbyte-ci --name=<connector-name> migrate_to_poetry`
It automatically does what you described below.

* `setup.py`
```python
from setuptools import find_packages, setup

MAIN_REQUIREMENTS = [

>     "airbyte-cdk~=0.1",

Reviewer 2: Let's call out that 0.1 here is an example. It's very dated, I want to make it clear.

    description="Source implementation for Openweather.",
    author="Airbyte",
    author_email="contact@airbyte.io",
    packages=find_packages(),
    install_requires=MAIN_REQUIREMENTS,
    package_data={
        "": [
            # Include yaml files in the package (if any)
            "*.yml",
            "*.yaml",
            # Include all json files in the package, up to 4 levels deep
            "*.json",
            "*/*.json",
            "*/*/*.json",
            "*/*/*/*.json",
            "*/*/*/*/*.json",
        ]
    },
    extras_require={
        "tests": TEST_REQUIREMENTS,
    },
)
```

* `pyproject.toml` 
```toml
[build-system]
requires = [ "poetry-core>=1.0.0",]
build-backend = "poetry.core.masonry.api"

[tool.poetry]
version = "0.2.1"
name = "source-openweather"
description = "Source implementation for Open Weather."
authors = [ "Airbyte <contact@airbyte.io>",]
license = "MIT"
readme = "README.md"
documentation = "https://docs.airbyte.com/integrations/sources/orb"
homepage = "https://airbyte.com"
repository = "https://github.com/airbytehq/airbyte"
[[tool.poetry.packages]]
include = "source_openweather"

[tool.poetry.dependencies]
python = "^3.9,<3.12"

> airbyte-cdk = "^0.74.0"

Reviewer 2: ```suggestion
airbyte-cdk = "^0"
```

Additionally, you should call out that the version of CDK should be unpinned unless you intentionally want to pin a specific one for the connector. In the future, all low-code sources will drop their pyproject.toml files and always run on latest CDK.

[tool.poetry.scripts]
source-openweather = "source_openweather.run:run"

[tool.poetry.group.dev.dependencies]

> pytest = "^6.2"

Reviewer 2: ```suggestion
pytest = "^8"
```

6.2 is ancient history, test deps should be updated.

> requests-mock = "^1.11.0"

Reviewer 2: ```suggestion
requests-mock = "*"
```

> pytest-mock = "^3.6.1"

Reviewer 2: ```suggestion
pytest-mock = "*"
```

```
4. Delete the `setup.py` file

> 5. Go to the connector folder and run `poetry lock` to generate the poetry file

Reviewer: The can also run `poetry check` to validate the package is correctly declared
